### PR TITLE
Fix how we split channel ids in channel queries

### PR DESCRIPTION
**Pull request by sstone**
We currently split in fixed-size chunks, which means that channel ids with the same block height may end up in different replies. Though this is not explicitly forbidden in the specs, it is implicitly required otherwise it becomes difficult to know when you've received all the replies you asked for and send another query.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 1]
```diff
@@ -1206,20 +1215,34 @@ object Router {
...
-          ShortChannelIdsChunk(firstBlock, numBlocks, group.toList)
+
+      // we use an iterator for efficiency
+      val it = shortChannelIds.iterator
+
+      // we want to split ids in different chunks, with the following rules by order of priority
+      // ids that have the same block height must be grouped in the same chunk
+      // chunk should contain `channelRangeChunkSize` ids
+      @tailrec
+      def loop(currentHeight: Int, currentChunk: List[ShortChannelId], acc: List[ShortChannelIdsChunk]): List[ShortChannelIdsChunk] = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]:
```suggestion
      def loop(currentChunk: List[ShortChannelId], acc: List[ShortChannelIdsChunk]): List[ShortChannelIdsChunk] = {
        val currentHeight = currentChunk.last.blockHeight
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 2]
```diff
@@ -158,7 +161,9 @@ case class RouteResponse(hops: Seq[ChannelHop], ignoreNodes: Set[PublicKey], ign
 // @formatter:off
 /** This is used when we get a TemporaryChannelFailure, to give time for the channel to recover (note that exclusions are directed) */
 case class ExcludeChannel(desc: ChannelDesc)
+
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: there are some formatting glitches

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Also why not use a `Queue` intead of prepend + reverse?

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 3]
```diff
@@ -127,4 +128,23 @@ class ChannelRangeQueriesSpec extends FunSuite {
     assert(Router.computeFlag(channels)(ef.shortChannelId, None, None, false) === (INCLUDE_CHANNEL_ANNOUNCEMENT | INCLUDE_CHANNEL_UPDATE_1 | INCLUDE_CHANNEL_UPDATE_2))
     assert(Router.computeFlag(channels)(ef.shortChannelId, None, None, true) === (INCLUDE_CHANNEL_ANNOUNCEMENT | INCLUDE_CHANNEL_UPDATE_1 | INCLUDE_CHANNEL_UPDATE_2 | INCLUDE_NODE_ANNOUNCEMENT_1 | INCLUDE_NODE_ANNOUNCEMENT_2))
   }
+
+  test("split short channel ids correctly ") {
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: Why not start with a very simple test with tiny chunk sizes?

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 4]
```diff
@@ -127,4 +128,23 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    def makeShortChannelIds(height: Int, count: Int): SortedSet[ShortChannelId] = SortedSet.empty[ShortChannelId] ++ (0 until count).map(c => ShortChannelId(height, c, 0))
+
+    val ids = makeShortChannelIds(42, 100) ++ makeShortChannelIds(43, 70) ++ makeShortChannelIds(44, 50) ++ makeShortChannelIds(45, 30) ++ makeShortChannelIds(50, 120)
+
+    // check that chunks do not overlap
+    def validate(chunks: List[ShortChannelIdsChunk]) : Boolean = chunks match {
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: this doesn't validate that we are not e.g. dropping or duplicating ids it seems

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 5]
```diff
@@ -127,4 +129,58 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    }
+
+    // all ids in different blocks, chunk size == 2
+    {
+      val ids = List(ShortChannelId(1000, 0, 0), ShortChannelId(1001, 0, 0), ShortChannelId(1002, 0, 0), ShortChannelId(1003, 0, 0), ShortChannelId(1004, 0, 0), ShortChannelId(1005, 0, 0))
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, 2)
+      assert(chunks == ShortChannelIdsChunk(1000, 2, List(ids(0), ids(1))) :: ShortChannelIdsChunk(1002, 2, List(ids(2), ids(3))) :: ShortChannelIdsChunk(1004, 2, List(ids(4), ids(5))) :: Nil)
+    }
+
+    // all ids in the same block
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: How about "all ids in same block, chunk size == 2"? This PR creates this very corner case we might as well test it?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 1 above)
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Because `Queue` is implemented with `List` so I don't think there would be a real gain here.

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala (hunk 5 above)
Comment (sstone) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: New extensive tests have been added

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 6]
```diff
@@ -1205,21 +1207,41 @@ object Router {
    * Have to split ids because otherwise message could be too big
    * there could be several reply_channel_range messages for a single query
    */
-  def split(shortChannelIds: SortedSet[ShortChannelId], channelRangeChunkSize: Int): List[ShortChannelIdsChunk] = {
-    // this algorithm can split blocks (meaning that we can in theory generate several replies with the same first_block/num_blocks
-    // and a different set of short_channel_ids) but it doesn't matter
-    if (shortChannelIds.isEmpty) {
-      List(ShortChannelIdsChunk(0, 0, List.empty))
+  def split(shortChannelIds: SortedSet[ShortChannelId], firstBlockNum: Long, numberOfBlocks: Long, channelRangeChunkSize: Int): List[ShortChannelIdsChunk] = {
+    val it = shortChannelIds.iterator.dropWhile(_.blockHeight < firstBlockNum).takeWhile(_.blockHeight < firstBlockNum + numberOfBlocks)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: I like including RFC snippets for that kind of precise things, saves time for reviewers.

```suggestion
    // BOLT 7: MUST encode a short_channel_id [...] in blocks first_blocknum to first_blocknum + number_of_blocks - 1
    val it = shortChannelIds.iterator.dropWhile(_.blockHeight < firstBlockNum).takeWhile(_.blockHeight < firstBlockNum + numberOfBlocks)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 7]
```diff
@@ -642,8 +642,10 @@ class Router(val nodeParams: NodeParams, watcher: ActorRef, initialized: Option[
           val shortChannelIds: SortedSet[ShortChannelId] = d.channels.keySet.filter(keep(firstBlockNum, numberOfBlocks, _))
           log.info("replying with {} items for range=({}, {})", shortChannelIds.size, firstBlockNum, numberOfBlocks)
           val chunks = Kamon.runWithSpan(Kamon.spanBuilder("split-channel-ids").start(), finishSpan = true) {
-            split(shortChannelIds, nodeParams.routerConf.channelRangeChunkSize)
+            split(shortChannelIds, firstBlockNum, numberOfBlocks, nodeParams.routerConf.channelRangeChunkSize)
           }
+
+          // we make sure that our replies cover the range specified in their queries
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: This comment seems out of place?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 8]
```diff
@@ -1014,7 +1016,7 @@ object Router {
    */
   def keep(firstBlockNum: Long, numberOfBlocks: Long, id: ShortChannelId): Boolean = {
     val TxCoordinates(height, _, _) = ShortChannelId.coordinates(id)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: nit: we could use the new `blockHeight` here instead of the whole `coordinates`

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 9]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+
+  def validate(chunk: ShortChannelIdsChunk) = {
+    require(!chunk.shortChannelIds.exists(!Router.keep(chunk.firstBlock, chunk.numBlocks, _)))
+  }
+
+  // check that chunks do not overlap and contain exactly the ids they were built from
+  def validate(ids: SortedSet[ShortChannelId], firstBlockNum: Long, numberOfBlocks: Long, chunks: List[ShortChannelIdsChunk]): Unit = {
+
+    @tailrec
+    def noOverloap(chunks: List[ShortChannelIdsChunk]): Boolean = chunks match {
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: nit:

```suggestion
    def noOverlap(chunks: List[ShortChannelIdsChunk]): Boolean = chunks match {
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 10]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+      case Nil => true
+      case a :: b :: _ if b.firstBlock < a.firstBlock + a.numBlocks => false
+      case _ => noOverloap(chunks.tail)
+    }
+
+    // aggregate ids from all chunks, to check that they match our input ids exactly
+    val chunkIds = SortedSet.empty[ShortChannelId] ++ chunks.map(_.shortChannelIds).flatten.toSet
+    val expected = ids.filter(Router.keep(firstBlockNum, numberOfBlocks, _))
+
+    if (expected.isEmpty) require(chunks == List(ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil)))
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: This looks redundant with the tests below

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 11]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+  test("split short channel ids correctly (basic tests") {
+    // no ids to split
+    {
+      val ids = Nil
+      val firstBlockNum = 10
+      val numberOfBlocks = 100
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, ids.size)
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil) :: Nil)
+    }
+    assert(Router.split(SortedSet.empty[ShortChannelId], 100, 1000, 100) == ShortChannelIdsChunk(100, 1000, Nil) :: Nil)
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: This is redundant with the test above?

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 12]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    val numberOfBlocks = 1000
+
+    validate(ids, firstBlockNum, numberOfBlocks, Router.split(ids, firstBlockNum, numberOfBlocks, 1))
+    validate(ids, firstBlockNum, numberOfBlocks, Router.split(ids, firstBlockNum, numberOfBlocks, 20))
+    validate(ids, firstBlockNum, numberOfBlocks, Router.split(ids, firstBlockNum, numberOfBlocks, 50))
+    validate(ids, firstBlockNum, numberOfBlocks, Router.split(ids, firstBlockNum, numberOfBlocks, 100))
+    validate(ids, firstBlockNum, numberOfBlocks, Router.split(ids, firstBlockNum, numberOfBlocks, 1000))
+  }
+
+  test("fuzzy test") {
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: It's not really a fuzzy test without randomness...

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 13]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    {
+      val ids = Nil
+      val firstBlockNum = 10
+      val numberOfBlocks = 100
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, ids.size)
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil) :: Nil)
+    }
+    assert(Router.split(SortedSet.empty[ShortChannelId], 100, 1000, 100) == ShortChannelIdsChunk(100, 1000, Nil) :: Nil)
+
+    // ids are all before the requested range
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]:
```suggestion
    // ids are all after the requested range
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 14]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    {
+      val ids = List(ShortChannelId(1000, 0, 0), ShortChannelId(1001, 0, 0), ShortChannelId(1002, 0, 0), ShortChannelId(1003, 0, 0), ShortChannelId(1004, 0, 0), ShortChannelId(1005, 0, 0))
+      val firstBlockNum = 10
+      val numberOfBlocks = 100
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, ids.size)
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil) :: Nil)
+    }
+
+    // ids are all after the requested range
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]:
```suggestion
    // ids are all before the requested range
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 15]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+
+    if (expected.isEmpty) require(chunks == List(ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil)))
+    chunks.foreach(validate)
+    require(chunks.head.firstBlock == firstBlockNum)
+    require(chunks.last.firstBlock + chunks.last.numBlocks == firstBlockNum + numberOfBlocks)
+    require(chunkIds == expected)
+    require(noOverloap(chunks))
+  }
+
+  test("split short channel ids correctly (basic tests") {
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: Defining this helper method:
```scala
def id(blockHeight: Int, txIndex: Int = 0, outputIndex: Int = 0) = ShortChannelId(blockHeight, txIndex, outputIndex)
```

allows for significant less verbosity

Comment (pm47) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: NB: particularly because we can remove the `txIndex`/`outputIndex` when they are not overriden.

Comment (pm47) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: By the way I think we are missing tests splitting on `outputIndex`?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 16]
```diff
@@ -1205,21 +1207,41 @@ object Router {
    * Have to split ids because otherwise message could be too big
    * there could be several reply_channel_range messages for a single query
    */
-  def split(shortChannelIds: SortedSet[ShortChannelId], channelRangeChunkSize: Int): List[ShortChannelIdsChunk] = {
-    // this algorithm can split blocks (meaning that we can in theory generate several replies with the same first_block/num_blocks
-    // and a different set of short_channel_ids) but it doesn't matter
-    if (shortChannelIds.isEmpty) {
-      List(ShortChannelIdsChunk(0, 0, List.empty))
+  def split(shortChannelIds: SortedSet[ShortChannelId], firstBlockNum: Long, numberOfBlocks: Long, channelRangeChunkSize: Int): List[ShortChannelIdsChunk] = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: I think there should be scaladoc for the args, in order to precisely know what this method does (particularly for `channelRangeChunkSize`).

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 17]
```diff
@@ -127,4 +129,163 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, ids) :: Nil)
+    }
+
+    // all ids in the same block, chunk size == 2
+    {
+      val ids = List(ShortChannelId(1000, 0, 0), ShortChannelId(1000, 1, 0), ShortChannelId(1000, 2, 0), ShortChannelId(1000, 3, 0), ShortChannelId(1000, 4, 0), ShortChannelId(1000, 5, 0))
+      val firstBlockNum = 900
+      val numberOfBlocks = 200
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, 2)
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, ids) :: Nil)
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: So, in this test, we are not respecting our own `channelRangeChunkSize` on purpose right?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 18]
```diff
@@ -1199,30 +1209,77 @@ object Router {
...
+
+      val first = it.next()
+      loop(first :: Nil, Nil)
     }
   }
 
+  /**
+   * Enforce max-size constraints for each chunk
+   * @param chunks list of short channel id chunks
+   * @param keepFirst flags that specifies wether to keep to first of last ids when there are too many of them
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: nit:

```suggestion
   * @param keepFirst flags that specifies whether to keep first of last ids when there are too many of them
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala (hunk 10 above)
Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: I still think this statement is unnecessary: the following lines re-test that (even for the case where `expected.isEmpty`)

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 19]
```diff
@@ -127,4 +131,226 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+      } else {
+        outputIndex = outputIndex + 1
+      }
+      output += ShortChannelId(height, txIndex, outputIndex)
+    }
+    output.toList
+  }
+
+  def validate(chunk: ShortChannelIdsChunk) = {
+    require(!chunk.shortChannelIds.exists(!Router.keep(chunk.firstBlock, chunk.numBlocks, _)))
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: nit: a bit more readable imo:

```suggestion
    require(chunk.shortChannelIds.forall(Router.keep(chunk.firstBlock, chunk.numBlocks, _)))
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 20]
```diff
@@ -127,4 +131,226 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+
+    @tailrec
+    def noOverlap(chunks: List[ShortChannelIdsChunk]): Boolean = chunks match {
+      case Nil => true
+      case a :: b :: _ if b.firstBlock < a.firstBlock + a.numBlocks => false
+      case _ => noOverlap(chunks.tail)
+    }
+
+    // aggregate ids from all chunks, to check that they match our input ids exactly
+    val chunkIds = SortedSet.empty[ShortChannelId] ++ chunks.map(_.shortChannelIds).flatten.toSet
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: nit:

```suggestion
    val chunkIds = SortedSet.empty[ShortChannelId] ++ chunks.flatMap(_.shortChannelIds).toSet
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 21]
```diff
@@ -127,4 +131,226 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    }
+
+    // all ids in the same block
+    {
+      val ids = makeShortChannelIds(1000, 100)
+      val firstBlockNum = 900
+      val numberOfBlocks = 200
+      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, 10)
+      assert(chunks == ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, ids) :: Nil)
+    }
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: I tried to add the following test and it doesn't pass. I think we should change the code to make it pass :)

```scala
    // don't start a new block in same chunk if it overflows chunk size
    {
      val ids = List(id(1000), id(1000, 1), id(1001), id(1001, 1))
      val firstBlockNum = 900
      val numberOfBlocks = 200
      val chunks = Router.split(SortedSet.empty[ShortChannelId] ++ ids, firstBlockNum, numberOfBlocks, 3)
      assert(chunks == ShortChannelIdsChunk(firstBlockNum, 100 + 1, List(ids(0), ids(1))) :: ShortChannelIdsChunk(1001, 1, List(ids(2), ids(3))) :: Nil)
    }
```

When we decide to start adding a new block to the current chunk, we should only do that if the whole block can fit in the chunk. Otherwise we should take the safer choice of starting a new chunk.
If we don't do that it's likely we'll run more often in the code that does the random split to avoid the 3200 elements limit which I think we should try to avoid (it's not really a problem to send a bit more chunks imo).

Comment (sstone) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: Yes the current "merge" algo is simplistic (but also very efficient). Being smarter would require to iterate twice and computing a `height -> count` map.
I propose instead to change the default `channel-range-chunk-size` to 1500 instead of 2500, it should reduces the probability of merging 2 chunks and ending up over the limit

Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: SGTM

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala (hunk 10 above)
Comment (sstone) [reply eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]: Yes but I'd like to keep it as intent here is really obvious

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 22]
```diff
@@ -1199,30 +1207,83 @@ object Router {
...
+     *
+     * @param maximumSize maximum size of the short channel ids list
+     * @return a chunk with at most `maximumSize` ids
+     */
+    def enforceMaximumSize(maximumSize: Int) = {
+      if (shortChannelIds.size <= maximumSize) this else {
+        // we use a random offset here, so even if shortChannelIds.size is much bigger than maximumSize (which should
+        // not happen) peers will eventually receive info about all channels in this chunk
+        val offset = Random.nextInt(shortChannelIds.size - maximumSize)
+        this.copy(shortChannelIds = this.shortChannelIds.drop(offset).take(maximumSize))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]:
```suggestion
        this.copy(shortChannelIds = this.shortChannelIds.slice(offset, offset + maximumSize))
```

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 23]
```diff
@@ -1199,30 +1207,83 @@ object Router {
...
+    /**
+     *
+     * @param maximumSize maximum size of the short channel ids list
+     * @return a chunk with at most `maximumSize` ids
+     */
+    def enforceMaximumSize(maximumSize: Int) = {
+      if (shortChannelIds.size <= maximumSize) this else {
+        // we use a random offset here, so even if shortChannelIds.size is much bigger than maximumSize (which should
+        // not happen) peers will eventually receive info about all channels in this chunk
+        val offset = Random.nextInt(shortChannelIds.size - maximumSize)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: off-by-one bug (reproduced with test provided below)

```suggestion
        val offset = Random.nextInt(shortChannelIds.size - maximumSize + 1)
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 24]
```diff
@@ -127,4 +131,215 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    val expected = ids.filter(Router.keep(firstBlockNum, numberOfBlocks, _))
+
+    if (expected.isEmpty) require(chunks == List(ShortChannelIdsChunk(firstBlockNum, numberOfBlocks, Nil)))
+    chunks.foreach(validate)
+    require(chunks.head.firstBlock == firstBlockNum)
+    require(chunks.last.firstBlock + chunks.last.numBlocks == firstBlockNum + numberOfBlocks)
+    require(chunkIds == expected)
+    require(noOverlap(chunks))
+  }
+
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]:
```suggestion

    test("limit channel ids chunk size") {
    val ids = makeShortChannelIds(1, 3)
    val chunk = ShortChannelIdsChunk(0, 10, ids)

    val res1 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(1).shortChannelIds
    assert(res1.toSet == Set(List(ids(0)), List(ids(1)), List(ids(2))))

    val res2 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(2).shortChannelIds
    assert(res2.toSet == Set(List(ids(0), ids(1)), List(ids(1), ids(2))))

    val res3 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(3).shortChannelIds
    assert(res3.toSet == Set(List(ids(0), ids(1), ids(2))))
  }
```

File: eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala [hunk 25]
```diff
@@ -127,4 +131,228 @@ class ChannelRangeQueriesSpec extends FunSuite {
...
+    val res1 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(1).shortChannelIds
+    assert(res1.toSet == Set(List(ids(0)), List(ids(1)), List(ids(2))))
+
+    val res2 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(2).shortChannelIds
+    assert(res2.toSet == Set(List(ids(0), ids(1)), List(ids(1), ids(2))))
+
+    val res3 = for (_ <- 0 until 100) yield chunk.enforceMaximumSize(3).shortChannelIds
+    assert(res3.toSet == Set(List(ids(0), ids(1), ids(2))))
+  }
+  test("split short channel ids correctly (basic tests") {
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/ChannelRangeQueriesSpec.scala]:
```suggestion

  test("split short channel ids correctly (basic tests") {
```
